Mitigate dancing nodes

davkal opened the pull request:
This is to mitigate the dancing node problem in #419 

Heuristics to implement towards iterative layout engine (minimize re-layouts):

* dont layout on edge removal
* dont layout on edge adding
* dont layout on node removal
* relayout on node adding
* track degradation of layouts via scoring (edge crossings, wrong edge direction) and only re-layout when exceeding a threshold

Be smart about node adding:
* track recent layouts for reuse of previous node locations of same node (instead of layout on node adding) 
* add node to 1st neighbour and stretch rest on same level

paulbellamy commented:
All seem reasonable. My only concern is that, while this will lead to fewer re-layouts, the re-layouts it does do will be bigger.  i.e. It will have added and removed a bunch of edges, removed some nodes, and not done a re-layout. Then when a new node comes along it will trigger a massive (confusing) re-layout, however the the degradataion scoring could probably help mitigate this.

IMO, degradation should be tracked strictly as a diff against the previous layout (or data-set). Tracking it as an objective score of edge-crossings, etc, could behave badly on layouts which are difficult to render (i.e. it will keep trying to lay it out, all with bad results). Would it be possible to compare the current layout's score, with the potential re-layout's score, and only apply the re-layout if the new score is "much higher"?

Also, IMO, (as mentioned on #419) it's not strictly an issue of nodes moving, as much as unpredictability.

davkal commented:
Simple heuristics are implemented. Once caveat with reusing old node locations is that they may overlap with locations of a different layout run. Reusing whole sets of old node locations wont get very far though because they only apply if the node sets are the same. Some postprocessing may be needed to spread out overlapping nodes.
Hard numbers: with a probe running on my mac for 10 minutes, 55 of 123 were layed out with the heuristic and prevented jumping. That's 44%.

davkal commented:
> All seem reasonable. My only concern is that, while this will lead to fewer re-layouts, the re-layouts it does do will be bigger. i.e. It will have added and removed a bunch of edges, removed some nodes, and not done a re-layout. Then when a new node comes along it will trigger a massive (confusing) re-layout, however the the degradataion scoring could probably help mitigate this.

There is no perceived correlation between the extent of node/edge changes and the subsequent layout change. A single edge can throw everything around.

> IMO, degradation should be tracked strictly as a diff against the previous layout (or data-set). Tracking it as an objective score of edge-crossings, etc, could behave badly on layouts which are difficult to render (i.e. it will keep trying to lay it out, all with bad results). Would it be possible to compare the current layout's score, with the potential re-layout's score, and only apply the re-layout if the new score is "much higher"?

Good idea with diffing the score against a re-layout.

tomwilkie commented:
I can't really follow the code I'm afraid; let me know if you want me to try harder.

I have tested on the dev env and it dramatically reduces dancing.  So massive +1 from me there.

Local testing shows there is some strange behaviour when selecting and unselecting nodes; the selected node doesn't always get centered properly, and the canvas doesn't always go back to where I'd expect.  Perhaps this is unrelated?

davkal commented on client/app/scripts/charts/__tests__/node-layout-test.js:
```diff
@@ -0,0 +1,238 @@
+    expect(nodes.n3.x).toBeLessThan(nodes.n4.x);
+    expect(nodes.n3.y).toEqual(nodes.n4.y);
+  });
+
```
Test exact positions against the first result.

davkal commented on client/app/scripts/charts/nodes-chart.js:
```diff
@@ -27,8 +29,9 @@ const NodesChart = React.createClass({
-      edges: {},
+      nodes: makeMap(),
+      edges: makeMap(),
+      history: makeList(),
```
Not used.

davkal commented on client/app/scripts/charts/nodes-chart.js:
```diff
@@ -166,7 +190,7 @@ const NodesChart = React.createClass({
+    const svgClassNames = this.state.maxNodesExceeded || nodeElements.size === 0 ? 'hide' : '';
 
     return (
       <div className="nodes-chart">
```
Replace `'` with a unicode character that doesnt mess up syntax highlighting.

tomwilkie commented:
Did a review on hangouts with david, and other than the comments he just added this LGTM.
